### DISCUSSION TITLE
Geometrically necessary dislocation density by dislocation type

### DISCUSSION THREAD

**Discussion by mfuhr-mpg** [mtex-toolbox/mtex#784] [Ask Anything]
Dear all,

I have a small question concerning the calculation of geometrically necessary dislocation density.
I am working with orientation maps of pure tungsten samples. Using the following code that is adopted from the MTEX homepage, I was able to calculate the "total" GND density:
```matlab
% Calculate grains
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',2*degree);
% Remove small grains
ebsd(grains(grains.grainSize<=5)) = [];

% Smooth orientation data
F = halfQuadraticFilter;
ebsd = smooth(ebsd('indexed'),F,'fill',grains);
ebsd = ebsd('indexed').gridify;
kappa = ebsd.curvature;
alpha = kappa.dislocationDensity;
dS = dislocationSystem.bcc(ebsd.CS);
nu = 0.28; %Poisson ratio for W

% Energy of the edge dislocations
dS(dS.isEdge).u = 1;

% Energy of the screw dislocations
dS(dS.isScrew).u = 1 - nu;

dSRot = ebsd.orientations * dS;
[rho,factor] = fitDislocationSystems(kappa,dSRot);

kappa = alpha.curvature;

%Calculate GND density
GND_density = factor*sum(abs(rho .* dSRot.u),2);

%Remove very small GND densities
GND_density(GND_density<1e5)=NaN;
```

Now, I am trying to separate the GND density using the type of dislocation, so that in the end I got two variables GND_density_screw and GND_density_edge that contain the geometrically necessary density of screw and edge dislocations, respectively.

Could you tell me how to implement my project using MTEX?

Thank you very much in advance!

Best regards
Maximilian Fuhr

**Comment by kilir** [comment]
Hi,
how about
```
GnD_screw= factor*sum(abs(rho(dSRot.isScrew) .* dSRot(dSRot.isScrew).u),2);
GnD_edge=factor*sum(abs(rho(dSRot.isEdge) .* dSRot(dSRot.isEdge).u),2);
GND_tot = factor*sum(abs(rho .* dSRot.u),2);
```
Cheers,
Rüdiger

**Comment by mfuhr-mpg** [comment]
Hi Rüdiger,

thanks for your quick response.
The code you suggested does indeed compile but it unfortunately does not yield the results I was looking for. The issue is that the size of the outputs does not fit the number of pixels of the initial orientation map. As an example: For a 500x476 pixel orientation map, the size of GnD_tot is 238000x1 (which fits the size of the orientation map), whereas the size of GnD_edge is 9153408x1 and of GnD_screw is 762784x1, respectively.
Do you have an idea on where the issue could come from and how we could solve it?
Thanks!

Best regards
Maximilian Fuhr

**Reply by kilir** [reply]
Hi Maximilian,
you are absolutely right. This one isn't so nice but seems to work for me:
```
GND_sS = abs(rho .* dSRot.u);

GND_E = GND_sS;
GND_E(dSRot.isScrew) = nan;

GND_S = GND_sS;
GND_S(dSRot.isEdge) = nan;

GND_E = factor*nansum(GND_E,2);
GND_S = factor*nansum(GND_S,2);

```
Cheers,
Rüdiger

**Reply by mfuhr-mpg** [reply]
Hi Rüdiger,

thanks, now it works perfectly!
It seemingly doesn't have to be nice if it works ;)

Best,
Max
